pytorch_lightning/utilities/apply_func.py

```diff
@@ -119,21 +119,33 @@ def apply_to_collection(
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
     if _is_dataclass_instance(data):
-        out_dict = {}
+        # make a deepcopy of the data,
+        # but do not deepcopy mapped fields since the computation would
+        # be wasted on values that likely get immediately overwritten
+        fields = {}
+        memo = {}
         for field in dataclasses.fields(data):
-            if field.init:
-                v = apply_to_collection(
-                    getattr(data, field.name),
-                    dtype,
-                    function,
-                    *args,
-                    wrong_dtype=wrong_dtype,
-                    include_none=include_none,
-                    **kwargs,
-                )
-                if include_none or v is not None:
-                    out_dict[field.name] = v
-        return elem_type(**out_dict)
+            field_value = getattr(data, field.name)
+            fields[field.name] = (field_value, field.init)
+            memo[id(field_value)] = field_value
```

Comment:
nit: everything passes without the memo. do you think you could write a test to ensure it works properly?

Reply:
The memo is meant to be an internal optimization that has no effect on the external API or observable functionality. However, there was one way in which this internal detail leaked: through non-init fields, which were deepcopied if the `memo` was omitted and passed through unchanged otherwise. The latest version rectifies this by ensuring that original non-init field values are explicitly set in the result.

With this change, I do not believe there is any reasonable test that could expose the usage of the memo.